Combined sets and functions

hypotext opened the pull request:
# Description

Related issues: https://github.com/penrose/TOG-2019/issues/11

* Merge (a subset of) real analysis and sets domains
* Improve support for nested predicates such as `Not(Onto())`
* Improve curve sampling 

# Implementation strategy and design decisions

- Some changes in the Substance module:
  - Separate types for sugared vs. core statements and their corresponding parser functions
  - Unified parser naming convention and code style
  - Removed legacy code
  - Reorganized functions in the order of the AST

# Examples with steps to reproduce them

* Discrete: `penrose functions-domain/spec-function.sub functions-domain/discrete.sty functions-domain/functions.dsl`
![image](https://user-images.githubusercontent.com/11740102/68352793-91a89500-00d5-11ea-881a-7a23795be987.png)
![image](https://user-images.githubusercontent.com/11740102/68352816-a8e78280-00d5-11ea-8c93-90a98498dcb5.png)
* Continuous: `penrose functions-domain/spec-function.sub functions-domain/continuous.sty functions-domain/functions.dsl`

| | onto | not onto|
| ------ | -----| -----|
|one-to-one | ![Screenshot 2019-11-02 18 12 36](https://user-images.githubusercontent.com/2762356/68077564-71f33300-fd9c-11e9-878e-067ed9e3c921.png) | ![Screenshot 2019-11-02 18 13 55](https://user-images.githubusercontent.com/2762356/68077575-8fc09800-fd9c-11e9-935a-1544ed29a5c6.png) |
|not one-to-one | ![image](https://user-images.githubusercontent.com/11740102/68068360-46cffb80-fd2a-11e9-92b5-84921ab85e6a.png) | ![image](https://user-images.githubusercontent.com/11740102/68068391-b2b26400-fd2a-11e9-93a1-b344f42ea145.png)|


# Checklist

- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran [Haddock](https://github.com/penrose/penrose/wiki/Writing-and-generating-documentation#generating-html-documentation-site) and there were no errors when generating the HTML site
- [x] My changes generate no new warnings
- [x] New and existing tests pass locally using `stack test`
- [x] My code follows the style guidelines of this project (e.g.: no HLint warnings)


# Open questions

Questions that require more discussion or to be addressed in future development:

- [ ] Alloy compatibility across machines
- [ ] Generating monotonic curves
- [ ] Sampling nice-looking colors

hypotext commented:
@wodeni Here are some screenshots of the four classes of functions with my revisions to the style, including the things I mentioned on Slack (see commit message). Feel free to try it out and make any changes, and then please go ahead and open a PR for this style!

![Screenshot 2019-11-01 20 13 40](https://user-images.githubusercontent.com/2762356/68063024-75c37e80-fce4-11e9-9b54-0e622b28638a.png)
![Screenshot 2019-11-01 20 12 23](https://user-images.githubusercontent.com/2762356/68063025-75c37e80-fce4-11e9-9306-baed744dd2ff.png)
![Screenshot 2019-11-01 20 11 42](https://user-images.githubusercontent.com/2762356/68063026-75c37e80-fce4-11e9-977a-c83e671e479d.png)
![Screenshot 2019-11-01 20 10 39](https://user-images.githubusercontent.com/2762356/68063027-765c1500-fce4-11e9-98ef-b618e3335391.png)

wodeni commented:
Following up the curve sampling problem @hypotext :

Original version looks like:
![Screenshot 2019-11-01 20 57 36](https://user-images.githubusercontent.com/11740102/68067007-4bd77f80-fd17-11e9-98a6-137254c227dd.png)


Seems that sampling y coordinates and use a linearly spaced vector of xs will give you reasonable results:
![image](https://user-images.githubusercontent.com/11740102/68066954-59403a00-fd16-11e9-9b05-196bd6a1594a.png)

Sometimes the curve goes out of the bbox. We can either have (1) smarter bbox computation or (2) stricter interpolation method. The current implementation is pretty loose and we could invest a lot of time to write a set of good algorithms. Note that D3 already does (2) (http://bl.ocks.org/d3indepth/b6d4845973089bc1012dec1674d3aff8), so my suggestion would be to defer interpolation until render-time unless necessary (if we ever want to optimize the curves themselves, then no!).

wodeni commented on src/Penrose/Shapes.hs:
```diff
@@ -322,30 +322,30 @@ type ComputedValue a = ([Property], [Value a] -> Value a)
 computedProperties ::
      (Autofloat a) => M.Map (ShapeTypeStr, Property) (ComputedValue a)
 computedProperties =
```
This was noted in #166 before.

wodeni replied:
Again, my guess is that the polymorphism forces the dict to be constructed repeatedly

hypotext replied:
We should really look into this... maybe time to create a minimal example? I still don't get why the function & objective (etc) dictionaries don't cause a huge polymorphism slowdown (unless they do and it doesn't show up in the profile)

wodeni replied:
Yeah, feel free to note this in one of the performance-related issues (e.g. #166 #213) and we can pick a time to look into it.

wodeni commented on react-renderer/src/Arrow.tsx:
```diff
@@ -23,13 +23,18 @@ class Arrow extends React.Component<IGPIPropsDraggable> {
-      (Math.cos(slope) * width) / 2,
-      (Math.sin(slope) * height) / 2
-    ];
+      const [offsetX, offsetY] = [0.0, 0.0];
```
This is causing the CI to fail. If the current solution doesn't work perfectly, consider reverting this change. We can look into more accurate arrow placement together soon.

hypotext commented:
Can you say more about why this would cause the CI to fail? Is it a failing
test or the “noUnusedLocals” config (which is not related to the code
correctness)? The new code does work pretty well (see the new diagrams)

On Mon, Nov 4, 2019 at 12:24 PM wodeni <notifications@github.com> wrote:

> *@wodeni* commented on this pull request.
> ------------------------------
>
> In react-renderer/src/Arrow.tsx
> <https://github.com/penrose/penrose/pull/278#discussion_r342172337>:
>
> >      const { width, height } = arrowheads[arrowheadStyle];
>      const slope = Math.atan2(ey - sy, ex - sx);
> -    const [offsetX, offsetY] = [
> -      (Math.cos(slope) * width) / 2,
> -      (Math.sin(slope) * height) / 2
> -    ];
> +      const [offsetX, offsetY] = [0.0, 0.0];
>
> This is causing the CI to fail. If the current solution doesn't work
> perfectly, consider reverting this change. We can look into more accurate
> arrow placement together soon.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/penrose/penrose/pull/278?email_source=notifications&email_token=AAVCM5CVZKMK76QBF4XIWKLQSBLFDA5CNFSM4JIBNH4KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCKGQIEQ#pullrequestreview-311231506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVCM5FWTINDLTF33MHCEB3QSBLFDANCNFSM4JIBNH4A>
> .
>

wodeni commented:
> Can you say more about why this would cause the CI to fail? Is it a failing test or the “noUnusedLocals” config (which is not related to the code correctness)? The new code does work pretty well (see the new diagrams)
> […](#)
> On Mon, Nov 4, 2019 at 12:24 PM wodeni ***@***.***> wrote: ***@***.**** commented on this pull request. ------------------------------ In react-renderer/src/Arrow.tsx <[#278 (comment)](https://github.com/penrose/penrose/pull/278#discussion_r342172337)>: > const { width, height } = arrowheads[arrowheadStyle]; const slope = Math.atan2(ey - sy, ex - sx); - const [offsetX, offsetY] = [ - (Math.cos(slope) * width) / 2, - (Math.sin(slope) * height) / 2 - ]; + const [offsetX, offsetY] = [0.0, 0.0]; This is causing the CI to fail. If the current solution doesn't work perfectly, consider reverting this change. We can look into more accurate arrow placement together soon. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub <#278?email_source=notifications&email_token=AAVCM5CVZKMK76QBF4XIWKLQSBLFDA5CNFSM4JIBNH4KYY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCKGQIEQ#pullrequestreview-311231506>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAVCM5FWTINDLTF33MHCEB3QSBLFDANCNFSM4JIBNH4A> .

Correct. The commented out code attempts to offset the marker size and place the arrowhead _exactly_ at the `(endX, endY)`. Thanks to this change, I went back and corrected the implementation (well, at least it looks more correct now). Here is the most current version on this branch, where the arrow is placed at the center of the label:

![image](https://user-images.githubusercontent.com/11740102/68157173-2915b880-ff1b-11e9-9e81-9a7b159ae514.png)

Correcting the offset computation and arrowhead definitions, here's the improved version (red dots drawn at the centers of the labels):
![image](https://user-images.githubusercontent.com/11740102/68162562-4b610380-ff26-11e9-94a4-9e1ca2078f6b.png)
Varying the stroke width and arrowhead size:
![image](https://user-images.githubusercontent.com/11740102/68162688-9c70f780-ff26-11e9-9355-337cabdcf43e.png)

hypotext commented:
Cool, the diagrams definitely look better now (here's one example), so thanks for making that change!

![image](https://user-images.githubusercontent.com/2762356/68167689-54f16800-ff34-11e9-8bb8-11ceedce2492.png)

wodeni commented on examples/elementLibrary.json:
```diff
@@ -16,5 +16,11 @@
     "uri": "real-analysis-domain/real-analysis.dsl"
+  },
+  {
+    "value": 3,
```
Resolve conflict later

hypotext commented on examples/linear-algebra-domain/linear-algebra.sty:
```diff
@@ -5,9 +5,15 @@
 -- Global constants and sizes
```
We're going to need to merge this file w/ Keenan's changes later

hypotext commented on examples/plugins/alloy/AlloyPlugin.java:
```diff
@@ -13,17 +13,16 @@
 
 
```
Would you be able to write a short description of how the Alloy plugin translates the Substance program into an Alloy description? This would be helpful for the paper.

hypotext commented on examples/set-theory-domain/venn-color.sty:
```diff
@@ -0,0 +1,99 @@
+-- TODO: Just optimize locations in 1D for now, and `r` proportional to `x` location
```
I think I am going to cherry-pick the color optimization commit onto another branch and revert it on this branch. We'll see what it does to the commit history

hypotext replied:
I just reverted it here and made another branch (https://github.com/penrose/penrose/tree/color-opt) w/o the revert commit. Not sure what will happen when `color-opt` gets merged--do we revert the revert commit? LMK if you know of a better way.

wodeni replied:
Not sure if I'm understanding correctly. I wouldn't mind keeping the changes, as long as you have it on another branch. Keeping the changes might actually be nicer for the merge in the future?

hypotext commented on src/Penrose/Functions.hs:
```diff
@@ -492,58 +489,48 @@ checkReturn (GPI v) _ = error "checkReturn: Computations cannot return GPIs"
-
-computeSurjection ::
-     Autofloat a => StdGen -> Integer -> Pt2 a -> Pt2 a -> ([Pt2 a], StdGen)
+sampleFunction [Val (IntV n), Val (FloatV xmin), Val (FloatV xmax), Val (FloatV ymin), Val (FloatV ymax), Val (StrV typ)] g 
```
I think it might be marginally faster to have these as separate functions (sampleSurjection, etc) so the runtime doesn't need to fall thru the case statement every time the function is evaluated. But it's probably not that much faster.

wodeni replied:
I'm not sure how the `Map` data structure in Haskell works. If the lookup cost scales with the size of the dict, there might be a trade-off between the two. Well, I'm personally inclined to have fewer more general comp functions. Let me know what you think.

hypotext commented on src/Penrose/Util.hs:
```diff
@@ -77,18 +77,27 @@ epsd = 10 ** (-10)
 -- General helper functions
 type Interval = (Float, Float) -- which causes type inference problems in Style for some reason.
 
+linspace :: (Autofloat a) => Int -> (a, a) -> [a]
```
I think there is an off-by-one error with this function where it oversamples by one increment. I'd recommend using `lerpN` instead. (Note that the "API" for `lerpN` is slightly different; the `n` there is the number of interior points, and it auto-includes the endpoints.) For example:

```
λ> linspace 2 (0, 1)                                                                                                                                                                                                                                                          
[0.0,1.0,2.0]                                                                                                                                                                                                                                                                 
λ> linspace 3 (0, 1)                                                                                                                                                                                                                                                          
[0.0,0.5,1.0,1.5]                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
λ> linspace 4 (0, 1)                                                                                                                                                                                                                                                          
[0.0,0.3333333333333333,0.6666666666666666,1.0,1.3333333333333333]                                                                                                                                                                                                            
λ> linspace 3 (-2, 2)                                                                                                                                                                                                                                                         
[-2.0,0.0,2.0,4.0]                                                                                                                                                                                                                                                            
λ> lerpN 0 1 2                                                                                                                                                                                                                                                                
[0.0,0.3333333333333333,0.6666666666666666,1.0]                                                                                                                                                                                                                               
λ> lerpN 0 1 1                                                                                                                                                                                                                                                                
[0.0,0.5,1.0]              
```

BTW, I think you already know how to load functions but just in case:

```
stack ghci [--profile]
:l Penrose.Util
```

wodeni replied:
Good catch. I'll use your function then.

hypotext commented on src/Penrose/Functions.hs:
```diff
@@ -815,8 +789,9 @@ sampleFunctionArea [GPI domain, GPI range, Val (FloatV xFrac), Val (FloatV yFrac
              pt_midleft = midpoint pt_bl pt_tl +: x_offset +: y_offset
-             right_curve = interpolateFn [pt_tr, pt_midright, pt_br]
-             left_curve = interpolateFn [pt_bl, pt_midleft, pt_tl]
+             -- FIXME: take k as arg
```
Did you want to come back and fix this now @wodeni?

wodeni replied:
Nope. Turns out that changing `k` doesn't fundamentally solve the shape-going-outta-bbox issue. So I didn't end up doing it. Removing the comment.

hypotext commented on src/Penrose/Functions.hs:
```diff
@@ -1292,10 +1302,37 @@ _centerArrow arr@("Arrow", _) s1@[x1, y1] s2@[x2, y2] [o1, o2] =
 -- Note: repel's energies are quite small so the function is scaled by repelWeight before being applied
 repel :: ObjFn
+
+-- TODO: factor line & arrow together OR account for the arrowhead
```
I know the copy-pasting is not good, but I didn't see any obvious simple way to let it fall through to the more general pattern matches (using an if statement would end the pattern match)

hypotext commented on src/Penrose/Functions.hs:
```diff
@@ -1403,7 +1440,7 @@ at :: ConstrFn
 
 lessThan :: ConstrFn
-lessThan [Val (FloatV x), Val (FloatV y)] = x - y
+lessThan [Val (FloatV x), Val (FloatV y)] = if x < y then 0 else (x - y)^2
```
This change should be unnecessary since the output `x` of any constraint energy is composed with `max(x, 0)^2` which has the same effect... unless you saw that it actually improved the layout? (Which it shouldn't)

wodeni replied:
Interestingly enough, it does! Here are examples from real analysis. My guess is that it's a unit issue. Since most other constraints are squared, I had to square this to make it "compatible" somehow?

`if` version:
![image](https://user-images.githubusercontent.com/11740102/68350135-6e79e780-00cd-11ea-88e0-b2656b7a0d38.png)
`x - y` version:
![image](https://user-images.githubusercontent.com/11740102/68350233-b436b000-00cd-11ea-9b21-d20558b27d5a.png)

hypotext replied:
Wow, that's drastic. Good to know.

hypotext commented on src/Penrose/Functions.hs:
```diff
@@ -1473,6 +1520,23 @@ contains [GPI rt@("Rectangle", _), GPI ar@("Arrow", _)] =
   in inRange startX lx rx + inRange startY ly ry + inRange endX lx rx +
      inRange endY ly ry
 
+contains [GPI sq@("Square", _), GPI ar@("Line", _)] =
```
Why did we need these constraints again? Do they work well? I'm surprised that adding a bunch of `inRange` constraints would work

wodeni replied:
I'm not sure if I was the one who wrote them. Checking the blame, it seems to be your change from 4 days ago?? https://github.com/penrose/penrose/blame/0eff99e63be0d75a3580f6cf505d5253aebac64a/src/Penrose/Functions.hs#L1520

hypotext commented:
Oh, and let's document this PR a bit before merging (though it looks like we'll be branching off into sets-functions again to finish the diagram iteration)

hypotext commented on examples/functions-domain/functions.dsl:
```diff
@@ -0,0 +1,28 @@
+predicate OneToOne : Map m
+predicate Onto : Map m
+predicate Bijection : Map m
+predicate PointIn : Set s * Point p
```
I noticed that the Alloy plugin generates `PointIn` statements. Can we generate `In` statements instead and remove `PointIn` from the Domain program? (`In` is more concise and has a more natural order of args)

wodeni replied:
Done.

wodeni commented:
> Oh, and let's document this PR a bit before merging (though it looks like we'll be branching off into sets-functions again to finish the diagram iteration)

What would you like to document?

wodeni commented:
Thanks for the comments @hypotext 🥇. I addressed them in the newest commits. Unless we want to make further changes regarding the unresolved discussions above, I'm happy to merge this branch (which includes a pretty big set of conflicts 😢 ).

hypotext commented:
Thanks for addressing the comments @wodeni! 

Ok, looks like we duplicated a bit of work on the style... let's coordinate better on that. Can you also check out issue #288 first?

hypotext commented:
Just added tiny bit more documentation to the PR, maybe you can write a bit about your big Substance parser refactor and then I think this is good to merge

wodeni commented:
Turns out that the polygonization doesn't work for all curves. @hypotext ran into this issue earlier when attempting to compute the polygon for a close curve with line segments. After resolving this issue, we can now appreciate the beautiful polygons:
![image](https://user-images.githubusercontent.com/11740102/68448753-86776700-01b2-11ea-991f-a057e43c6dd1.png)

hypotext commented:
Put that in the Louvre! :)

On Thu, Nov 7, 2019 at 11:01 PM wodeni <notifications@github.com> wrote:

> Turns out that the polygonization doesn't work for all curves. @hypotext
> <https://github.com/hypotext> ran into this issue earlier when attempting
> to compute the polygon for a close curve with line segments. After
> resolving this issue, we can now appreciate the beautiful polygons:
> [image: image]
> <https://user-images.githubusercontent.com/11740102/68448753-86776700-01b2-11ea-991f-a057e43c6dd1.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/penrose/penrose/pull/278?email_source=notifications&email_token=AAVCM5F5MHHKG2CB7OSYOP3QSTQCBA5CNFSM4JIBNH4KYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEDOVBEY#issuecomment-551374995>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAVCM5AHBLAFWV5FFFVE2YDQSTQCBANCNFSM4JIBNH4A>
> .
>
